src/ads.jl

```diff
@@ -0,0 +1,574 @@
+    parent::Union{Nothing, ADSTaylorSolution{T, N, M}},
+    left::Union{Nothing, ADSTaylorSolution{T, N, M}},
+    right::Union{Nothing, ADSTaylorSolution{T, N, M}}) where {T, N, M} =
+    ADSTaylorSolution{T, N, M}(depth, t, lo, hi, x, p, parent, left, right)
+
+# 3-arg constructor
+function ADSTaylorSolution(lo::AbstractVector{T}, hi::AbstractVector{T},
+    x::AbstractVector{TaylorN{T}}) where {T}
+    @assert length(lo) == length(hi)
+    N = length(lo)
+    M = length(x)
+    return ADSTaylorSolution(0, zero(T), SVector{N, T}(lo), SVector{N, T}(hi),
+    SVector{M, TaylorN{T}}(x), nothing, nothing, nothing, nothing)
+end
+
+## Custom print
+
+function Base.show(io::IO, n::ADSTaylorSolution{T, N, M}) where {T, N, M}
+    s = Vector{String}(undef, N)
+    for i in eachindex(s)
+        s[i] = string(SVector{2, T}(n.lo[i], n.hi[i]))
+    end
+    plural = M > 1 ? "s" : ""
+    print(io, "t: ", n.t, " s: ", join(s, "×"), " x: ", M, " ",
+        TaylorN{T}, " variable" * plural)
+end
+
+## AbstractTrees API
+
+# Left (right) child constructors
+function leftchild!(
+    parent::ADSTaylorSolution{T, N, M}, t::T, lo::SVector{N, T},
+    hi::SVector{N, T}, x::SVector{M, TaylorN{T}},
+    p::Union{Nothing, SVector{M, Taylor1{TaylorN{T}}}} = nothing,
+    left::Union{Nothing, ADSTaylorSolution{T, N, M}} = nothing,
+    right::Union{Nothing, ADSTaylorSolution{T, N, M}} = nothing) where {T, N, M}
+    isnothing(parent.left) || error("Left child is already assigned")
+    node = ADSTaylorSolution{T, N, M}(parent.depth + 1, t, lo, hi, x, p, parent,
+                                      left, right)
+    parent.left = node
+end
+
+function leftchild!(parent::ADSTaylorSolution{T, N, M},
+    node::ADSTaylorSolution{T, N, M}) where {T, N, M}
+    # set `node` as left child of `parent`
+    parent.left = node
+end
+
+function rightchild!(
+    parent::ADSTaylorSolution{T, N, M}, t::T, lo::SVector{N, T},
+    hi::SVector{N, T}, x::SVector{M, TaylorN{T}},
+    p::Union{Nothing, SVector{M, Taylor1{TaylorN{T}}}} = nothing,
+    left::Union{Nothing, ADSTaylorSolution{T, N, M}} = nothing,
+    right::Union{Nothing, ADSTaylorSolution{T, N, M}} = nothing) where {T, N, M}
+    isnothing(parent.right) || error("Right child is already assigned")
+    node = ADSTaylorSolution{T, N, M}(parent.depth + 1, t, lo, hi, x, p, parent,
+                                      left, right)
+    parent.right = node
+end
+
+function rightchild!(parent::ADSTaylorSolution{T, N, M},
+    node::ADSTaylorSolution{T, N, M}) where {T, N, M}
+    # set `node` as right child of `parent`
+    parent.right = node
+end
+
+# AbstractTrees interface
+function AbstractTrees.children(node::ADSTaylorSolution)
+    if isnothing(node.left) && isnothing(node.right)
+        ()
+    elseif isnothing(node.left) && !isnothing(node.right)
+        (node.right,)
+    elseif !isnothing(node.left) && isnothing(node.right)
+        (node.left,)
+    else
+        (node.left, node.right)
```

Comment:
This function is used to iterate a tree, in this case `ADSTaylorSolution`, so we cannot return a `::Nothing` since `nothing` cannot be converted to `ADSTaylorSolution`.

Reply:
But then we do have some type unstability. My point is to somehow avoid it.

Reply:
Just thought that it's perhaps worth mentioning that the current proposal is following both the AbstractTrees.jl [interface](https://juliacollections.github.io/AbstractTrees.jl/stable/#The-Abstract-Tree-Interface) and AFAIK the implementation here follows closely the binary tree example provided in AbstractTrees.jl itself (see e.g. [this method](https://github.com/JuliaCollections/AbstractTrees.jl/blob/af3fabd36631efe361bc8e1ba45f9b84e898b459/test/examples/binarytree.jl#L31). This is only to say that as it is now `children` is in principle as type unstable as the example provided by them in their documentation. Maybe the compiler is somehow able to resolve this safely? @LuEdRaMo what is the output of `@code_warntype children(...)` called on, say, a leaf node and a non-leaf node?

Reply:
The output of `@code_warntype` is the same for all nodes:
```julia
Arguments
  #self#::Core.Const(AbstractTrees.children)
  node::ADSTaylorSolution{Float64, 2, 4}
Body::Tuple{Vararg{Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}}}
1 ── %1  = Base.getproperty(node, :left)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %2  = TaylorIntegration.isnothing(%1)::Bool
└───       goto #4 if not %2
2 ── %4  = Base.getproperty(node, :right)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %5  = TaylorIntegration.isnothing(%4)::Bool
└───       goto #4 if not %5
3 ── %7  = ()::Core.Const(())
└───       return %7
4 ┄─ %9  = Base.getproperty(node, :left)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %10 = TaylorIntegration.isnothing(%9)::Bool
└───       goto #7 if not %10
5 ── %12 = Base.getproperty(node, :right)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %13 = TaylorIntegration.isnothing(%12)::Bool
│    %14 = !%13::Bool
└───       goto #7 if not %14
6 ── %16 = Base.getproperty(node, :right)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %17 = Core.tuple(%16)::Tuple{Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}}
└───       return %17
7 ┄─ %19 = Base.getproperty(node, :left)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %20 = TaylorIntegration.isnothing(%19)::Bool
│    %21 = !%20::Bool
└───       goto #10 if not %21
8 ── %23 = Base.getproperty(node, :right)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %24 = TaylorIntegration.isnothing(%23)::Bool
└───       goto #10 if not %24
9 ── %26 = Base.getproperty(node, :left)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %27 = Core.tuple(%26)::Tuple{Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}}
└───       return %27
10 ┄ %29 = Base.getproperty(node, :left)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %30 = Base.getproperty(node, :right)::Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}
│    %31 = Core.tuple(%29, %30)::Tuple{Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}, Union{Nothing, ADSTaylorSolution{Float64, 2, 4}}}
└───       return %31
```

Reply:
AFAIK the `AbstractTrees` API does not require `children` to return a `Tuple`, so I tried returning a `Vector{ADSTaylorSolution{T, N, M}}`, it works and is type stable.

Reply:
If the returned tuple yields no `@code_warning` problems, I think we can go with it; otherwise, returning a `Vector` seems the way to go!

Thanks a lot for taking a look into this!

Reply:
I think I would go for the tuple, otherwise, the performance we gain from type stability we might lose by allocating a vector each time children is called. Thanks for checking this!